<?xml version="1.0"?>
<robot name="test_model" xmlns:xacro="http://www.ros.org/wiki/xacro">

  <xacro:include filename="/home/alpaca/m_robot_frame/src/mrobot_frame/urdf/xacro/tracker.xacro"/>
  <xacro:include filename="/home/alpaca/m_robot_frame/src/mrobot_frame/urdf/xacro/depthcam.xacro"/>

  <link name="base_footprint"/>

  <joint name="footprint_joint" type="fixed">
    <parent link="base_footprint"/>
    <child link="base_link"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
  </joint>

  <link
    name="base_link">
    <inertial>
      <origin
        xyz="-0.00327575378179794 1.43828480751708E-07 0.0382570233322481"
        rpy="0 0 0" />
      <mass
        value="0.12172971186255" />
      <inertia
        ixx="0.000155535446390653"
        ixy="-5.36352216135528E-10"
        ixz="7.8671141829149E-06"
        iyy="0.000348311182872133"
        iyz="2.84330475668362E-10"
        izz="0.000448529037289916" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>

      </geometry>
      <material
        name="">
        <color
          rgba="0.5 0.5 0.5 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>

      </geometry>
    </collision>
  </link>
  <link
    name="left_wheel">
    <inertial>
      <origin
        xyz="5.16864329114242E-13 -0.00250604866328853 -1.39067576898633E-13"
        rpy="0 0 0" />
      <mass
        value="0.0240789296710246" />
      <inertia
        ixx="6.09971814850164E-06"
        ixy="6.45944123912712E-17"
        ixz="-3.44144697240244E-17"
        iyy="1.06370873786235E-05"
        iyz="-1.79734526046358E-17"
        izz="6.09971814832533E-06" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>

      </geometry>
      <material
        name="">
        <color
          rgba="0.0 0.0 0.0 1.0" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>

      </geometry>
    </collision>
  </link>
  <joint
    name="left_back_wheel"
    type="continuous">
    <origin
      xyz="-0.0417945961781014 0.0426173263101681 0.0143"
      rpy="0 0 0" />
    <parent
      link="base_link" />
    <child
      link="left_wheel" />
    <axis
      xyz="0 1 0" />
  </joint>
  <link
    name="right_wheel">
    <inertial>
      <origin
        xyz="5.46958311975487E-13 0.00119738550817131 -1.28340046923192E-13"
        rpy="0 0 0" />
      <mass
        value="0.0240789296708149" />
      <inertia
        ixx="6.09971814847366E-06"
        ixy="-7.34575071316395E-17"
        ixz="-5.6543885633186E-17"
        iyy="1.06370873785448E-05"
        iyz="1.96556533378466E-17"
        izz="6.09971814826249E-06" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>

      </geometry>
      <material
        name="">
        <color
          rgba="0.0 0.0 0.0 1.0" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>

      </geometry>
    </collision>
  </link>
  <joint
    name="right_back_wheel"
    type="continuous">
    <origin
      xyz="-0.0417945961781014 -0.0413086631550841 0.0143"
      rpy="0 0 0" />
    <parent
      link="base_link" />
    <child
      link="right_wheel" />
    <axis
      xyz="0 1 0" />
  </joint>
  <link
    name="lidar_Link">
    <inertial>
      <origin
        xyz="1.93246810444315E-34 1.63232502450634E-18 -0.005"
        rpy="0 0 0" />
      <mass
        value="0.0125663706143592" />
      <inertia
        ixx="1.36135681655558E-06"
        ixy="-3.1861838222649E-54"
        ixz="2.42841104008608E-39"
        iyy="1.36135681655558E-06"
        iyz="-3.95610362478956E-38"
        izz="2.51327412287184E-06" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>

      </geometry>
      <material
        name="">
        <color
          rgba="0.0 0.0 0.0 1.0" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>

      </geometry>
    </collision>
  </link>

  <joint
    name="lidar"
    type="fixed">
    <origin
      xyz="-0.00679459617810136 0 0.07"
      rpy="0 0 0" />
    <parent
      link="base_link" />
    <child
      link="lidar_Link" />
    <axis
      xyz="0 0 0" />
  </joint>
  
  <gazebo reference="lidar_Link">
		<sensor type="ray" name="rplidar">
			<pose>
				0 0 0 0 0 0
			</pose>
			<visualize>
				true
			</visualize>
			<update_rate>
				5.5
			</update_rate>
			<ray>
				<scan>
					<horizontal>
						<samples>
							360
						</samples>
						<resolution>
							1
						</resolution>
						<min_angle>
							-3.14159
						</min_angle>
						<max_angle>
							3.14159
						</max_angle>
					</horizontal>
				</scan>
				<range>
					<min>
						0.10
					</min>
					<max>
						30.0
					</max>
					<resolution>
						0.01
					</resolution>
				</range>
			</ray>
			<plugin name="gazebo_rplidar" filename="libgazebo_ros_laser.so">
				<frameName>
					lidar_Link
				</frameName>
				<topicName>
					/scan
				</topicName>
			</plugin>
		</sensor>
	</gazebo>

  <link
    name="front_wheel">
    <inertial>
      <origin
        xyz="-0.00408064642229678 -0.0062927199823444 0.000800505948184976"
        rpy="0 0 0" />
      <mass
        value="0.00353049982552195" />
      <inertia
        ixx="1.436470020138E-07"
        ixy="1.60888341860673E-08"
        ixz="-6.90898403180046E-24"
        iyy="1.58024273920057E-07"
        iyz="8.97456348025334E-24"
        izz="1.29348226016588E-07" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>

      </geometry>
      <material
        name="">
        <color
          rgba="0.41176 0.41176 0.41176 1.0" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>

      </geometry>
    </collision>
  </link>
  <joint
    name="front_wheel"
    type="fixed">
    <origin
      xyz="0.0592860502441954 0.00651011920048181 -0.00041210325440747"
      rpy="0 0 0" />
    <parent
      link="base_link" />
    <child
      link="front_wheel" />
    <axis
      xyz="0 0 0" />
  </joint>

  <!-- <link name="base_link">
    <visual>
      <origin rpy="0 0 0 " xyz="0 0 0.005"/>
      <geometry>
          <box size="0.4 0.4 0.01"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0 " xyz="0 0 0.005"/>
      <geometry>
        <box size="0.4 0.4 0.01"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0 " xyz="0 0 0.005"/>
      <geometry>
        <box size="0.4 0.4 0.01"/>
      </geometry>
      <mass value="100"/>
      <inertia ixx="1.0265" ixy="0" ixz="0" iyy="1.3944999999999999" iyz="0" izz="2.1320000000000001"/>
    </inertial>
  </link> -->

  <!-- TEST -->

  <xacro:realsense_d435 sensor_name="d435" parent_link="base_link" rate="10">
    <origin rpy="0 0 0 " xyz="-0.0134539119794152 -0.000496123411225748 0.112"/>
  </xacro:realsense_d435>
  <gazebo reference="d435_bottom_screw_frame">  
    <sensor type="depth" name="camera">
      <always_on>true</always_on>
      <update_rate>20.0</update_rate>
      <camera>
        <image>
          <format>R8G8B8</format>
          <width>640</width>
          <height>480</height>
        </image>
        <clip>
          <near>0.05</near>
          <far>8.0</far>
        </clip>
      </camera>
      <plugin name="kinect_camera_controller" filename="libgazebo_ros_openni_kinect.so">
        <cameraName>camera</cameraName>
        <alwaysOn>true</alwaysOn>
        <updateRate>10</updateRate>
        <imageTopicName>rgb/image_raw</imageTopicName>
        <depthImageTopicName>depth/image_raw</depthImageTopicName>
        <pointCloudTopicName>depth/points</pointCloudTopicName>
        <cameraInfoTopicName>rgb/camera_info</cameraInfoTopicName>
        <depthImageCameraInfoTopicName>depth/camera_info</depthImageCameraInfoTopicName>
        <frameName>camera</frameName>
        <baseline>0.1</baseline>
        <distortion_k1>0.0</distortion_k1>
        <distortion_k2>0.0</distortion_k2>
        <distortion_k3>0.0</distortion_k3>
        <distortion_t1>0.0</distortion_t1>
        <distortion_t2>0.0</distortion_t2>
        <pointCloudCutoff>0.4</pointCloudCutoff>
      </plugin>
    </sensor>
  </gazebo> 

  <!-- <xacro:realsense_T265 sensor_name="camera" parent_link="base_link" rate="30.0">
    <origin rpy="0 0 0" xyz="0 0 0.5"/>
  </xacro:realsense_T265> -->

  <!-- <xacro:realsense_R200 sensor_name="camera" parent_link="base_link" rate="30.0">
    <origin rpy="0 0 0" xyz="0 0 0.5"/>
  </xacro:realsense_R200> -->


</robot>
